File: searches/binary_search.cpp
#include <algorithm>
#include <iostream>
#include <vector>

/**
 * @brief Binary Search implementation in C++
 *
 * @param arr Sorted vector
 * @param target Value to search for
 * @return int Index of target or -1 if not found
 */
int binary_search(const std::vector<int> &arr, int target) {
  int left = 0;
  int right = arr.size() - 1;

  while (left <= right) {
    int mid = left + (right - left) / 2;

    if (arr[mid] == target) {
      return mid;
    }
    if (arr[mid] < target) {
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }
  return -1;
}

int main() {
  std::vector<int> arr = {2, 3, 4, 10, 40};
  int target = 10;

  std::cout << "Array: ";
  for (int i : arr)
    std::cout << i << " ";
  std::cout << "\nTarget: " << target << "\n";

  int result = binary_search(arr, target);

  if (result != -1)
    std::cout << "Element found at index " << result << std::endl;
  else
    std::cout << "Element not found" << std::endl;

  return 0;
}

File: searches/binary_search.rs
/**
 * Binary Search implementation in Rust
 */

pub fn binary_search<T: Ord>(arr: &[T], target: &T) -> Option<usize> {
    let mut left = 0;
    let mut right = arr.len(); // right is exclusive in Rust ranges usually, but here we use indexes

    while left < right {
        let mid = left + (right - left) / 2;

        if &arr[mid] == target {
            return Some(mid);
        }
        if &arr[mid] < target {
            left = mid + 1;
        } else {
            right = mid;
        }
    }
    None
}

fn main() {
    let arr = vec![2, 3, 4, 10, 40];
    let target = 10;
    
    match binary_search(&arr, &target) {
        Some(index) => println!("Element found at index {}", index),
        None => println!("Element not found"),
    }
}
